graph-http/src/types/nextlink.rs

```diff
@@ -0,0 +1,8 @@
+pub struct NextLink;
+
+#[derive(Deserialize)]
+pub struct NextLinkValues<V> {
+    pub(crate) value: Vec<V>,
+    #[serde(rename = "@odata.nextLink")]
+    pub(crate) next_link: Option<String>,
+}
```

Comment:
This is what I was thinking as far as a trait goes using the previous `json` method changes I suggested. With this you will have to remove `next_link()` method from `ODataLink` replace all calls of `ODataLink` where` next_link()` is used with `ODataNextLink` but thats fine. 

```suggestion
pub trait ODataNextLink {
    fn next_link(&self) -> Option<String>;
}

impl ODataNextLink for serde_json::Value {
    fn next_link(&self) -> Option<String> {
        self["@odata.nextLink"].as_str().map(|s| s.to_string())
    }
}
```

Reply:
Ok, I understand now. There was some quiproquo between us on what `V` was. 
I thought you meant the singular item (ex `User`) but you meant a struct containing a Vec<> of items (ex `Users`). That's why I was confused about the paging. 
Well, here goes my english proficiency.

The problem with this solution is that we still require the caller to declare a struct and implement the trait manually,  that's why we went the macro way. I find it cumbersome to have a bunch of `User`, `Users`, `impl ODataNextLink for Users`, etc for each struct that I fetch from graph_api when I just want to get every `User` in my tenant and do my business logic on it. 

I feel like if we want to return a `Vec<Users>`  we should declare ourselves the `Users` (with a more generic name ofc) struct with all the possible attributes and methods (ODataLink ? ) to save the caller the trouble. And then implement a method so the caller can flatten this `Vec<Users>` to `Vec<User>`.

```rust
impl<'a> IntoResponseAsync<'a, NextLink> {
    pub async fn json<V>(self) -> GraphResult<GraphResponse<Vec<NextLinkValues<V>>>>
        where
                for<'de> V: serde::Deserialize<'de>,
    {
        if self.error.is_some() {
            return Err(self.error.unwrap_or_default());
        }
        let mut values = vec![];
        let request = self.client.build().await;
        let response = request.send().await?;
        let headers = response.headers().clone();
        let status = response.status();
        let url = GraphUrl::from(response.url());
        let body: NextLinkValues<V> = response.json().await.map_err(GraphFailure::from)?;
        let mut next_link = body.next_link.clone().and_then(|url| GraphUrl::parse(&url).ok());
        values.push(body);
        while let Some(url) = next_link {
            self.client.set_url(url);
            let request = self.client.build().await;
            let response = request.send().await?;
            let body: NextLinkValues<V> = response.json().await.map_err(GraphFailure::from)?;
            next_link = body.next_link.clone().and_then(|url| GraphUrl::parse(&url).ok());
            values.push(body);
        }
        Ok(GraphResponse::new(url, values, status, headers))
    }
}

#[derive(Deserialize)]
pub struct NextLinkValues<V> {
    pub value: Vec<V>,
    #[serde(rename = "@odata.nextLink")]
    pub next_link: Option<String>,
}

pub trait Foo<V> {
    fn to_v(self) -> Vec<V>;
}

impl<V> Foo<V> for Vec<NextLinkValues<V>> {
    fn to_v(self) -> Vec<V> {
        let mut b = vec![];

        for mut c in self.into_iter().map(|a| a.value) {
            b.append(&mut c)
        }
        b
    }
}
```

so the caller can do something like that :
![image](https://user-images.githubusercontent.com/2280811/194527887-403c6a9b-8efb-43ef-bdef-30d77952804c.png)

Reply:
> There was some quiproquo between us on what V was.

I have no idea what `quiproquo` means 😆. If you are saying quid pro quo, which is an actual English saying, then that is not what it means 😄.  But its fine I think the word to use there should just be miscommunication.

There was some miscommunication between us on what V was.

**Quid pro quo** just means a favor or advantage granted or expected in return for something

Whereas **miscommunication** means failure to communicate adequately.

I am honestly not trying to be rude or mean by explaining those words. I genuinely hope that it helps you.

> The problem with this solution is that we still require the caller to declare a struct and implement the trait manually, that's why we went the macro way. I find it cumbersome to have a bunch of User, Users, impl ODataNextLink for Users, etc for each struct that I fetch from graph_api when I just want to get every User in my tenant and do my business logic on it.'

I can see where your coming from and I get it. But I also know that this is something where people have different opinions on what they want and every use case is different. Some people do want to declare a User and Users struct and have a way to implement traits such as this. This is why I mentioned that your solution is great for those who just want to get all of the users regardless. But that doesn't take into account other use cases. 

As a library maintainer my goal is to provide the ability to do things the way each person wants. There are libraries that take a very opinionated approach to things but I typically have tried not to.

> I feel like if we want to return a Vec<Users> we should declare ourselves the Users (with a more generic name ofc) struct with all the possible attributes and methods (ODataLink ? ) to save the caller the trouble. And then implement a method so the caller can flatten this Vec<Users> to Vec<User>.

I have no problem with us flattening the `Vec<Users>` to just a `Vec<User>`. My example wasn't meant to say that the `Vec` should be a `Vec<Users>`. I was just trying to explain with an example what I meant about the trait. 👍 


Ok with all of that said, I think your probably fine staying with what you have now. Except maybe lets clean up some of the code a bit like you have in your reply there. Meaning lets use the `while let loop`. You don't need to do a to_v method. You can just return the vec flattened completley. 

And I dont know if this was intentional, but you can just use the `flatten()` method on an iterator of `Vec<Vec<T>>` in order to turn it into a `Vec<T>`.

Reply:
> > There was some quiproquo between us on what V was.
> 
> I have no idea what `quiproquo` means 😆. If you are saying quid pro quo, which is an actual English saying, then that is not what it means 😄. But its fine I think the word to use there should just be miscommunication.
> 
> There was some miscommunication between us on what V was.
> 
> **Quid pro quo** just means a favor or advantage granted or expected in return for something
> 
> Whereas **miscommunication** means failure to communicate adequately.
> 
> I am honestly not trying to be rude or mean by explaining those words. I genuinely hope that it helps you.

Sorry, I got confused with the french word [quiproquo](https://www.linguee.com/french-english/translation/quiproquo.html). Thanks for the explanations !

> Ok with all of that said, I think your probably fine staying with what you have now. Except maybe lets clean up some of the code a bit like you have in your reply there. Meaning lets use the `while let loop`. You don't need to do a to_v method. You can just return the vec flattened completley.

Sure. I'll do that.

> And I dont know if this was intentional, but you can just use the `flatten()` method on an iterator of `Vec<Vec<T>>` in order to turn it into a `Vec<T>`. Again im not sure if you knew that already or if there was some other reason for the method you wrote in your reply.
> 
> An example of flatten:
> 
> ```rust
>     let mut vec1: Vec<Vec<&str>> = vec![
>         vec!["a", "b", "c"],
>         vec!["e", "f", "g"]
>     ];
> 
>     let vec2: Vec<&str> = vec1.into_iter().flatten().collect();
>     println!("{vec2:?}"); // -> ["a", "b", "c", "e", "f", "g"]
> ```

Huh. I did know about it. I remember trying it before the `map` but not after. 
It is indeed much cleaner: 
``` rust
impl<V> Foo<V> for Vec<NextLinkValues<V>> {
    fn to_v(self) -> Vec<V> {
        self.into_iter().map(|a| a.value).flatten().collect()
    }
}
```